### PR TITLE
Support different AMP & buffer configurations in one experiment, fix minor bugs

### PR DESCRIPTION
**Pull request by justheuristic**
New features:
- CollaborativeOptimizer can now combine fp16=True and reuse_grad_buffers=True with a special scaler
- CollaborativeOptimizer peers with reuse_grad_buffers=True and reuse_grad_buffers=False can now co-exist
- CollaborativeOptimizer peers with and without AMP can now co-exist

The new behavior of CollaborativeOptimizer with fp16 is:
* grad_scaler=None: regular fp32 behavior
* reuse_grad_buffers=False with GradScaler: works as usual, independently un-scales each tensor before accumulation, does not affect internal optimizer
* reuse_grad_buffers=True with GradScaler: when calling scaler.step(opt), it will raise error and complain that it requires HivemindGradScaler
* reuse_grad_buffers=False with HivemindGradScaler: applies unscale/update only around global optimizer step

### CODE REVIEW

File: hivemind/optim/collaborative.py [hunk 1]
```diff
@@ -344,38 +370,36 @@ def accumulated_grads(self) -> Iterator[torch.Tensor]:
...
+            # user is responsible for accumulating gradients in .grad buffers
+            assert batch_size == self.batch_size_per_step, "Custom batch size is not implemented for reuse_grad_buffers"
+        else:
+            alpha = float(batch_size) / self.batch_size_per_step
+            for grad_buf, grad_acc in zip(self._grad_buffers(), self.accumulated_grads()):
+                grad_acc.add_(grad_buf.to(grad_acc.device), alpha=alpha)
 
     @torch.no_grad()
     def apply_accumulated_grads_(self, scale_by: Optional[float] = None):
-        if self.reuse_grad_buffers:
```
Comment (justheuristic) [inline hivemind/optim/collaborative.py:365]: This previously caused a bug where reuse=True peers would not be scaled by `scale_by`. As a result, they would have larger gradients and dominate the reuse=False peers.

File: hivemind/optim/collaborative.py [hunk 2]
```diff
@@ -344,38 +370,36 @@ def accumulated_grads(self) -> Iterator[torch.Tensor]:
         """local gradient accumulators"""
         if self.reuse_grad_buffers:
             yield from self._grad_buffers()
-        elif self._grads is None:
-            with torch.no_grad():
-                self._grads = [
-                    torch.zeros_like(grad, device=self.accumulate_grads_on) for grad in self._grad_buffers()
-                ]
-        yield from self._grads
```
Comment (justheuristic) [inline hivemind/optim/collaborative.py:352]: This wold actually an error with reuse_grad_buffers=True, but it worked because noone asked for more than len(grad_buffers) elements

File: hivemind/optim/collaborative.py [hunk 3]
```diff
@@ -197,6 +200,8 @@ def load_state_from_peers(self, **kwargs):
                 try:
                     self.averager.load_state_from_peers(timeout=self.load_state_timeout, **kwargs)
                     break
+                except KeyboardInterrupt:
+                    raise
```
Comment (justheuristic) [inline hivemind/optim/collaborative.py:204]: This would previously cause a deadlock if user would KeyboardInterrupt CollaborativeOptimizer during load_state_from_peers

File: hivemind/optim/grad_scaler.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        assert isinstance(optimizer, DecentralizedOptimizerBase)
+        if self._is_running_global_step:
+            super().unscale_(optimizer.opt)
+            return True
+        else:
+            self._check_inf_per_device(optimizer.opt)
+            self._optimizer_states_to_reset.add(id(optimizer))
+            return False
+
+    def step(self, optimizer, *args, **kwargs):
```
Comment (borzunov) [inline hivemind/optim/grad_scaler.py:41]: nit:

```suggestion
    def step(self, optimizer: Optimizer, *args, **kwargs) -> bool:
```

File: hivemind/optim/grad_scaler.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        else:
+            for opt_id in self._optimizer_states_to_reset:
+                self._per_optimizer_states[opt_id] = _refresh_per_optimizer_state()
+            self._optimizer_states_to_reset.clear()
+            return False
+
+    def _unscale_grads_(
+        self, optimizer: Optimizer, inv_scale: torch.Tensor, found_inf: torch.Tensor, allow_fp16: bool
+    ) -> Dict[torch.device, torch.Tensor]:
+        return super()._unscale_grads_(optimizer, inv_scale, found_inf, allow_fp16=True)
```
Comment (borzunov) [inline hivemind/optim/grad_scaler.py:72]: Why does it ignore the `allow_fp16` value always setting it to True?

File: hivemind/optim/collaborative.py (hunk 3 above)
Comment (borzunov) [reply hivemind/optim/collaborative.py:204]: Maybe replace `except BaseException` -> `except Exception` instead of this change?

File: hivemind/optim/collaborative.py [hunk 6]
```diff
@@ -251,6 +266,10 @@ def step(self, batch_size: Optional[int] = None, **kwargs):
 
             # divide accumulators by local steps to recover the true average grad w.r.t. local_samples_accumulated
             self.apply_accumulated_grads_(scale_by=1.0 / self.local_updates_accumulated)
+            if grad_scaler:
```
Comment (borzunov) [inline hivemind/optim/collaborative.py:269]: nit:
```suggestion
            if grad_scaler is not None:
```

File: hivemind/optim/collaborative.py [hunk 7]
```diff
@@ -279,13 +298,21 @@ def step(self, batch_size: Optional[int] = None, **kwargs):
                     f"Skipped averaging: collaboration consists of " f"{self.collaboration_state.num_peers} peer(s).",
                 )
 
-            self.opt.step()
+            if grad_scaler:
```
Comment (borzunov) [inline hivemind/optim/collaborative.py:301]: nit:
```suggestion
            if grad_scaler is not None:
```

File: hivemind/optim/collaborative.py [hunk 8]
```diff
@@ -279,13 +298,21 @@ def step(self, batch_size: Optional[int] = None, **kwargs):
...
+            else:
+                self.opt.step()
+
             self.reset_accumulated_grads_()
             self.local_samples_accumulated = self.local_updates_accumulated = 0
             self.collaboration_state.register_step(current_step + 1)
             self.averager.local_step = current_step + 1
             self.collaboration_state_updated.set()
             self.update_scheduler()
+            if grad_scaler:
```
Comment (borzunov) [inline hivemind/optim/collaborative.py:313]: nit:
```suggestion
            if grad_scaler is not None:
```

File: hivemind/optim/grad_scaler.py [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    def are_grads_finite(self, optimizer: DecentralizedOptimizerBase):
```
Comment (borzunov) [inline hivemind/optim/grad_scaler.py:74]:
```suggestion
    def are_grads_finite(self, optimizer: DecentralizedOptimizerBase) -> bool:
```

File: hivemind/optim/grad_scaler.py [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+                super().step(optimizer.opt, *args, **kwargs)
+            else:
+                logger.warning("Skipping global step due to gradient over/underflow")
+            return True
+        else:
+            super().step(optimizer)
+            self._optimizer_states_to_reset.add(optimizer)
+            return False
+
+    def update(self, new_scale=None):
```
Comment (borzunov) [inline hivemind/optim/grad_scaler.py:54]: nit:

```suggestion
    def update(self, new_scale: Optional[float] = None) -> bool:
```

**Review by borzunov — CHANGES_REQUESTED** [review]

File: hivemind/optim/grad_scaler.py (hunk 5 above)
Comment (justheuristic) [reply hivemind/optim/grad_scaler.py:72]: that's the same trick that fairscale uses to allow training without master fp32 weights
https://github.com/facebookresearch/fairscale/blob/main/fairscale/optim/grad_scaler.py
(got referred there by @TimDettmers)

Added a quick comment explaining that

File: hivemind/optim/collaborative.py [hunk 11]
```diff
@@ -344,38 +371,38 @@ def accumulated_grads(self) -> Iterator[torch.Tensor]:
         """local gradient accumulators"""
         if self.reuse_grad_buffers:
             yield from self._grad_buffers()
-        elif self._grads is None:
-            with torch.no_grad():
+            return
+        else:
```
[the remark this answers is not in the document]
Comment (justheuristic) [reply hivemind/optim/collaborative.py:375]: I respectfully disagree here: it does reduce indentation, but:
- the code in the else clause is quite simple as it is
- crucially, last time the code was wrong because i've accidentally removed return and didn't think about the ramifications

If you do insist, please state that explicitly, i'll remove the else clause anyway

File: hivemind/optim/collaborative.py (hunk 3 above)
Comment (justheuristic) [reply hivemind/optim/collaborative.py:204]: Sometime in future, yes, but currently i'm afraid that it may accidentally forward CancelledError or something like that.

File: hivemind/optim/collaborative.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (borzunov) [reply hivemind/optim/collaborative.py:375]: In this case, I do not insist :)

[the remark this answers is not in the document]
Comment (borzunov) [reply hivemind/optim/collaborative.py:375]: I would say that `else:` after `return` is obviously redundant and seems like a bug, so I insist on choosing one of the two options:

1. `return` without `else:` (preferred since it implies less indentation, more clarity)
2. `else:` without `return`

File: hivemind/optim/collaborative.py (hunk 3 above)
Comment (borzunov) [reply hivemind/optim/collaborative.py:204]: Okay

**Review by borzunov — APPROVED** [review]
